### PR TITLE
Add type and field checks to `MeshSeq`

### LINKED ISSUE #169: Add checks in `MeshSeq.form`
First mentioned in https://github.com/mesh-adaptation/goalie/pull/137/files#r1520802694, I'd like to add type and field checks in `MeshSeq.form`, to catch up with other methods that are more developed. Currently we only check that `get_form` returns a dictionary and we do it in `indicate_errors`.

### PR DESCRIPTION
**Pull request by ddundo**
Closes #169.

This PR:
- adds `MeshSeq._outputs_consistent` and calls it in `__init__`. The method checks that function spaces, initial conditions, and forms are given in a dictionary format with `MeshSeq.fields` as keys
- adds unit tests for above
- tidies up some other tests

The only problem is that now `solid_body_rotation_split.py` raises a "missing field" error because it considers one field at a time. ~~Easy solution is to remove the missing field check. What do you think @jwallwork23 @acse-ej321? :)~~
I removed the test here as it will be removed by #176 (will merge this later).

### CODE REVIEW

File: test/test_mesh_seq.py [hunk 1]
```diff
@@ -43,26 +44,73 @@ def test_inconsistent_dim(self):
...
+            )
+            with self.assertRaises(AssertionError) as cm:
+                mesh_seq.initial_condition()
+        elif function_name == "form":
+
+            def get_form(mesh_seq):
+                def form(index, solutions):
+                    return {"new_field": 1.0}
+
+                return form
```
Comment (ddundo) [inline test/test_mesh_seq.py:78]: I will rewrite this with lambda functions for conciseness.

**Comment by ddundo** [comment]
Converted to a draft - I want to rethink this a bit

**Comment by ddundo** [comment]
@jwallwork23 @acse-ej321 ready for review! Sorry for the confusion. 

I think this is a much better solution now since the checks are only done once and they are done in `MeshSeq.__init__`, so any errors will be caught immediately. Currently errors in `get_form` would be caught at the very end of `indicate_errors`, after all forward and adjoint solves - so it's a significant waste of time.

**Review by jwallwork23 — APPROVED** [review]
Nice! Thanks for this.

**Pull request merged**